Kernel/Devices/GPU/3dfx/Definitions.h

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2023, Edwin Rijkee <edwin@virtualparadise.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Array.h>
+#include <AK/Types.h>
+
+namespace Kernel::VoodooGraphics {
+
+enum class VGAPort : u16 {
+    AttributeController = 0x3c0,
+    MiscOutputWrite = 0x3c2,
+    SequencerIndex = 0x3c4,
+    SequencerData = 0x3c5,
+    MiscOutputRead = 0x3cc,
+    GraphicsControllerIndex = 0x3ce,
+    GraphicsControllerData = 0x3cf,
+    CrtcIndex = 0x3d4,
+    CrtcData = 0x3d5,
+    InputStatus1 = 0x3da,
+};
+
+enum Flags {
+    // CR3: CRTC Horizontal Blanking End
+    CR3_CompatibilityRead = 1 << 7,
+
+    // CR11: CRTC Vertical Sync End
+    CR11_EnableVertInt = 1 << 5,
+    CR11_CRTCRegsWriteProt = 1 << 7,
+
+    // CR17 CRTC Mode Control
+    CR17_TimingEnable = 1 << 7,
+    CR17_ByteWordMode = 1 << 6,
+
+    // GR6: Graphics Controller Miscellaneous
+    GR6_MemoryMapEGAVGAExtended = 1 << 2,
+
+    // AR10: Attribute Controller Mode
+    AR10_GraphicsMode = 1 << 0,
+    AR10_PixelWidth = 1 << 6,
+
+    // SR0: Sequencer Reset
+    SR0_AsynchronousReset = 1 << 0,
+    SR0_SynchronousReset = 1 << 1,
+
+    // SR1: Sequencer Clocking Mode
+    SR1_DotClock8 = 1 << 0,
+
+    // Miscellaneous Output
+    MISCOUTPUT_CRTCAddressColor = 1 << 0,
+    MISCOUTPUT_ClockSelectPLL = 0b1100,
+    MISCOUTPUT_VerticalSyncPositive = 1 << 7,
+    MISCOUTPUT_HorizontalSyncPositive = 1 << 6,
+
+    // dacMode
+    DACMODE_2X = 1 << 0,
+
+    // vgaInit0
+    VGAINIT0_FIFODepth8Bit = 1 << 2,
+    VGAINIT0_EnableVgaExtensions = 1 << 6,
+    VGAINIT0_WakeUpSelect3C3 = 1 << 8,
+    VGAINIT0_EnableAltReadback = 1 << 10,
+    VGAINIT0_ExtendedShiftOut = 1 << 12,
+
+    // vidProcCfg: Video Processor Configuration
+    VIDPROCCFG_VideoProcessorEnable = 1 << 0,
+    VIDPROCCFG_DesktopSurfaceEnable = 1 << 7,
+    VIDPROCCFG_DesktopCLUTBypass = 1 << 10,
+    VIDPROCCFG_DesktopPixelFormat32Bit = 0b11 << 18,
+    VIDPROCCFG_2XMode = 1 << 26,
+};
+
+struct PLLSettings {
+    static const i32 reference_frequency_in_khz = 14318;
+    i32 m = 0;
+    i32 n = 0;
+    i32 k = 0;
+
+    int frequency_in_khz() const
+    {
+        return (reference_frequency_in_khz * (n + 2) / (m + 2)) >> k;
+    }
+
+    u32 register_value() const
+    {
+        return (n << 8) | (m << 2) | k;
+    }
+};
+
+struct ModeRegisters {
+    u32 vid_screen_size = 0;
+    u32 vid_desktop_overlay_stride = 0;
+    u8 misc_out_reg = 0;
+    u32 vga_init0 = 0;
+    u32 vid_proc_cfg = 0;
+    u32 dac_mode = 0;
+    u32 pll_ctrl0 = 0;
+    Array<u8, 0x1c> cr = { 0 }; // CRT Controller Registers
+    Array<u8, 0x09> gr = { 0 }; // Graphics Controller Registers
+    Array<u8, 0x15> ar = { 0 }; // Attribute Controller Registers
+    Array<u8, 0x05> sr = { 0 }; // Sequencer Registers
```

Comment:
On Andrews comment:
Maybe make these their own structs with a size assertion

Reply:
I think I would prefer to only do it how @ADKaster suggested and keep it as an array, because it will allow to keep the part where the registers are actually written simpler.